### PR TITLE
option to add taints per nodepool

### PR DESCRIPTION
**Pull request by DStorck**
old method was to apply taints at cluster creation via nodeConfig (essentially type), now taints can be added per nodePool

fixes https://github.com/samsung-cnct/k2/issues/680

### CODE REVIEW

File: ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.6/units.kubelet.part.jinja2 [hunk 1]
```diff
@@ -43,6 +43,7 @@
       --api_servers=https://apiserver.{{ cluster.name }}.internal \
       --register-node=true \
       {% if node.nodeConfig.taints is defined %}--register-with-taints={% set comma = joiner(",") %}{% for taint in node.nodeConfig.taints  %}{{ comma() }}{{taint.key}}={{taint.value}}:{{taint.effect}}{% endfor %}{% endif %} \
+      {% if node.taintConfig is defined %}--register-with-taints={% set comma = joiner(",") %}{% for taint in node.taintConfig.taints  %}{{ comma() }}{{taint.key}}={{taint.value}}:{{taint.effect}}{% endfor %}{% endif %} \
```
Comment (coffeepac) [inline ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.6/units.kubelet.part.jinja2]: can we specify multiple `--register-with-taints` to the kubelet or do we need to join these two lists?

File: ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.7/units.kubelet.part.jinja2 [hunk 2]
```diff
@@ -43,6 +43,7 @@
       --api_servers=https://apiserver.{{ cluster.name }}.internal \
       --register-node=true \
       {% if node.nodeConfig.taints is defined %}--register-with-taints={% set comma = joiner(",") %}{% for taint in node.nodeConfig.taints  %}{{ comma() }}{{taint.key}}={{taint.value}}:{{taint.effect}}{% endfor %}{% endif %} \
+      {% if node.taintConfig is defined %}--register-with-taints={% set comma = joiner(",") %}{% for taint in node.taintConfig.taints  %}{{ comma() }}{{taint.key}}={{taint.value}}:{{taint.effect}}{% endfor %}{% endif %} \
```
Comment (coffeepac) [inline ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.7/units.kubelet.part.jinja2]: same as above, I don't think we can specify this switch twice but I could be wrong!

File: ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.6/units.kubelet.part.jinja2 (hunk 1 above)
Comment (DStorck) [inline ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.6/units.kubelet.part.jinja2]: it has been working just fine, but I'm okay to rework it if we think it is cleaner to just to it once

**Comment by DStorck** [comment]
@davidewatson Thanks for the reminder on definitions.json - that is done. 
@coffeepac Your comment as you know was dealt with and now we have schedulingConfig.

File: schemas/config/v1/definitions.json [hunk 3]
```diff
@@ -113,6 +113,14 @@
       },
       "minItems": 1,
       "type": "array"
+    },
+    "schedulingConfig": {
```
Comment (davidewatson) [inline schemas/config/v1/definitions.json]: This should be plural.

File: schemas/config/v1/schedulingConfig.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "id": "http://judkins.house/apis/k2/v1/schedulingConfig.json",
+  "$$target": "schedulingConfig.json",
+  "title": "Scheduling configuration",
+  "description": "Describes configuration of taints added per nodePool",
+
+  "properties": {
+    "taints": { "$ref": "kubeNodeTaint.json" }
```
Comment (davidewatson) [inline schemas/config/v1/schedulingConfig.json]: Should this be an array of taints?

File: schemas/config/v1/schedulingConfig.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "id": "http://judkins.house/apis/k2/v1/schedulingConfig.json",
+  "$$target": "schedulingConfig.json",
+  "title": "Scheduling configuration",
+  "description": "Describes configuration of taints added per nodePool",
+
+  "properties": {
```
Comment (davidewatson) [inline schemas/config/v1/schedulingConfig.json]: Configs objects should have a name and kind to facilitate extensions.

File: schemas/config/v1/schedulingConfig.json (hunk 4 above)
Comment (DStorck) [reply schemas/config/v1/schedulingConfig.json]: our plan was to have this also be the home of 'labels' in the future

Comment (davidewatson) [reply schemas/config/v1/schedulingConfig.json]: According to your documentation, this is an array:

```
+  schedulingConfig:
 +    - &defaultScheduling
 +      taints:
 +        - key: testKey
 +          value: testValue
 +          effect: NoSchedule
```

File: schemas/config/v1/schedulingConfig.json (hunk 5 above)
Comment (DStorck) [reply schemas/config/v1/schedulingConfig.json]: i put them in @davidewatson , let me know if that is what you were thinking.

File: schemas/config/v1/schedulingConfig.json [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      "description": "Scheduling configuration for taints and labels.",
+      "enum": ["scheduling"],
+      "type": "string"
+    },
+    "taints": {
+      "description": "List of taints to apply to node",
+      "items": {
+        "$ref": "kubeNodeTaint.json"
+      },
+      "type": "array"
```
Comment (davidewatson) [inline schemas/config/v1/schedulingConfig.json]: This line needs to be swapped with line above. `type` belongs in the `taints` stanza.

Comment (DStorck) [reply schemas/config/v1/schedulingConfig.json]: are you sure? putting it a line above would make it part of the `items`

Comment (davidewatson) [reply schemas/config/v1/schedulingConfig.json]: You're right! I misread this CI error: ```An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ValueError: Expecting object: line 26 column 2 (char 727)```

**Pull request merged**